File: src/main/kotlin/community/flock/localpipeline/provider/trello/TrelloProvider.kt
package community.flock.localpipeline.provider.trello

import community.flock.localpipeline.model.WorkItem
import community.flock.localpipeline.provider.WorkItemProvider
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.get
import io.ktor.client.request.parameter

class TrelloProvider(
    private val client: HttpClient,
    private val apiKey: String,
    private val token: String,
) : WorkItemProvider {

    override val name = "Trello"

    override suspend fun fetchAssignedItems(): List<WorkItem> {
        val member: TrelloMember = client.get("https://api.trello.com/1/members/me") {
            parameter("key", apiKey)
            parameter("token", token)
        }.body()

        val cards: List<TrelloCard> = client.get("https://api.trello.com/1/members/${member.id}/cards") {
            parameter("key", apiKey)
            parameter("token", token)
            parameter("fields", "id,name,desc,shortUrl,idList,labels,idBoard")
        }.body()

        if (cards.isEmpty()) return emptyList()

        // Collect unique board IDs and fetch board names + lists
        val boardIds = cards.map { it.idBoard }.distinct()
        val boardNames = mutableMapOf<String, String>()
        val listNames = mutableMapOf<String, String>()

        for (boardId in boardIds) {
            val board: TrelloBoard = client.get("https://api.trello.com/1/boards/$boardId") {
                parameter("key", apiKey)
                parameter("token", token)
                parameter("fields", "id,name")
            }.body()
            boardNames[boardId] = board.name

            val lists: List<TrelloList> = client.get("https://api.trello.com/1/boards/$boardId/lists") {
                parameter("key", apiKey)
                parameter("token", token)
                parameter("fields", "id,name")
            }.body()
            lists.forEach { listNames[it.id] = it.name }
        }

        return cards.map { card ->
            WorkItem(
                id = card.id.take(8),
                title = card.name,
                description = card.desc?.takeIf { it.isNotBlank() }?.take(500),
                status = listNames[card.idList],
                priority = null,
                labels = card.labels.mapNotNull { it.name.takeIf { n -> n.isNotBlank() } },
                source = "Trello",
                team = boardNames[card.idBoard],
                url = card.shortUrl,
            )
        }
    }
}

File: src/providers/trello/trello-provider.ts
import type { WorkItem } from "../../model/work-item.js";
import type { WorkItemProvider, Board } from "../provider.js";
import type { TrelloMember, TrelloCard, TrelloBoard, TrelloList } from "./trello-types.js";

export class TrelloProvider implements WorkItemProvider {
  name = "Trello";
  private boardFilter?: string;

  constructor(
    private apiKey: string,
    private token: string,
  ) {}

  setBoardFilter(boardId: string): void {
    this.boardFilter = boardId;
  }

  async fetchBoards(): Promise<Board[]> {
    const member = await this.get<TrelloMember>("/members/me");
    const boards = await this.get<TrelloBoard[]>(`/members/${member.id}/boards`, {
      fields: "id,name",
      filter: "open",
    });
    return boards.map((b) => ({ id: b.id, name: b.name }));
  }

  private params(): URLSearchParams {
    return new URLSearchParams({ key: this.apiKey, token: this.token });
  }

  private async get<T>(path: string, extra?: Record<string, string>): Promise<T> {
    const params = this.params();
    if (extra) {
      for (const [k, v] of Object.entries(extra)) params.set(k, v);
    }
    const res = await fetch(`https://api.trello.com/1${path}?${params}`);
    if (!res.ok) {
      throw new Error(`Trello API error: ${res.status} ${res.statusText}`);
    }
    return res.json();
  }

  async addComment(itemId: string, comment: string): Promise<void> {
    const params = this.params();
    params.set("text", comment);
    const res = await fetch(`https://api.trello.com/1/cards/${itemId}/actions/comments?${params}`, {
      method: "POST",
    });
    if (!res.ok) {
      throw new Error(`Trello API error: ${res.status} ${res.statusText}`);
    }
  }

  async markDone(item: WorkItem): Promise<void> {
    const card = await this.get<TrelloCard>(`/cards/${item.id}`, { fields: "idBoard,idList" });
    const lists = await this.get<TrelloList[]>(`/boards/${card.idBoard}/lists`, { fields: "id,name" });
    const doneList = lists.find((l) => l.name.toLowerCase() === "done");
    if (!doneList) {
      throw new Error("No 'Done' list found on this board");
    }
    const params = this.params();
    params.set("idList", doneList.id);
    const res = await fetch(`https://api.trello.com/1/cards/${item.id}?${params}`, {
      method: "PUT",
    });
    if (!res.ok) {
      throw new Error(`Trello API error: ${res.status} ${res.statusText}`);
    }
  }

  async fetchAssignedItems(): Promise<WorkItem[]> {
    const member = await this.get<TrelloMember>("/members/me");
    const allCards = await this.get<TrelloCard[]>(`/members/${member.id}/cards`, {
      fields: "id,name,desc,shortUrl,idList,labels,idBoard",
    });

    const filteredByBoard = this.boardFilter
      ? allCards.filter((c) => c.idBoard === this.boardFilter)
      : allCards;

    if (filteredByBoard.length === 0) return [];

    const boardIds = [...new Set(filteredByBoard.map((c) => c.idBoard))];
    const boardNames = new Map<string, string>();
    const listNames = new Map<string, string>();

    await Promise.all(
      boardIds.map(async (boardId) => {
        const [board, lists] = await Promise.all([
          this.get<TrelloBoard>(`/boards/${boardId}`, { fields: "id,name" }),
          this.get<TrelloList[]>(`/boards/${boardId}/lists`, { fields: "id,name" }),
        ]);
        boardNames.set(boardId, board.name);
        for (const list of lists) {
          listNames.set(list.id, list.name);
        }
      }),
    );

    const excludedLists = new Set(["done", "in review"]);
    const cards = filteredByBoard.filter((c) => {
      const listName = listNames.get(c.idList)?.toLowerCase();
      return !listName || !excludedLists.has(listName);
    });

    return cards.map((card) => ({
      id: card.id.slice(0, 8),
      title: card.name,
      description: card.desc?.trim() ? card.desc.slice(0, 500) : undefined,
      status: listNames.get(card.idList),
      priority: undefined,
      labels: card.labels.map((l) => l.name).filter((n) => n.length > 0),
      source: "Trello",
      team: boardNames.get(card.idBoard),
      url: card.shortUrl,
    }));
  }
}
